File: CC/Trees and Graph/buildorder.js
/*
	@param Array of nodes : graph
	@return : Array of node : any building order 
*/

function DirectedGraphNode(val){
	this.val = val;
	this.neighbors = [];
}
// BFS 
var buildOrder1 = function(graph){
	// build parents count map 
	var map = new Map();
	for(var i=0; i<graph.length; i++){
		var curNode = graph[i];
		for(var j=0; j<curNode.neighbors.length; j++){
			if(map.has(curNode.neighbors[j])){
				map.set(curNode.neighbors[j], map.get(curNode.neighbors[j])+1);
			}else{
				map.set(curNode.neighbors[j], 1);
			}
		}
	}

	// find start point 
	var queue = [], result = [];
	for(var c=0; c<graph.length; c++){
		if(map.has(graph[c])){
			queue.push(graph[c]);
			result.push(graph[c]);
		}
	}

	// go for topological sort 
	while(queue.size()>0){
		var cur = queue.shift();
		for(var t=0; t<cur.neighbors.length; t++){
			map.set(cur.neighbors[t], map.get(cur.neighbors[t])-1);
			if(map.get(cur.neighbors[t])===0){
				queue.push(cur.neighbors[t]);
				result.push(cur.neighbors[t]);
			}
		}
	}
	return result;
};

// DFS 
var buildOrder2 = function(graph){
	// build parents count map 
	var map = new Map(), result = [];
	for(var i=0; i<graph.length; i++){
		var curNode = graph[i];
		for(var j=0; j<curNode.neighbors.length; j++){
			if(map.has(curNode.neighbors[j])){
				map.set(curNode.neighbors[j], map.get(curNode.neighbors[j])+1);
			}else{
				map.set(curNode.neighbors[j], 1);
			}
		}
	}
	// topological sort
	for(var k=0; k<graph.length; k++){
		if(!map.has(graph[k])) dfs(map, result, graph[k]);
	}
	return result;
};


function dfs (map, result, curNode){
	result.push(curNode);
	for(var i=0; i<curNode.neighbors.length; i++){
		map.set(curNode.neighbors[i], map.get(curNode.neighbors[i])-1);
		if(map.get(curNode.neighbors[i])===0) dfs(map, result, curNode.neighbors[i]);
	}
}

File: CC/Trees and Graph/BuildOrder.java
/**
 * Definition for Directed graph.
 * class DirectedGraphNode {
 *     int label;
 *     ArrayList<DirectedGraphNode> neighbors;
 *     DirectedGraphNode(int x) { label = x; neighbors = new ArrayList<DirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */    
    public ArrayList<DirectedGraphNode> topSort(ArrayList<DirectedGraphNode> graph) {
        // write your code here
        // build the map of parent count 
        HashMap<DirectedGraphNode, Integer> map = new HashMap<DirectedGraphNode, Integer> ();
        for(DirectedGraphNode node : graph){
            for(DirectedGraphNode neighbor : node.neighbors){
                if(map.containsKey(neighbor)){
                    map.put(neighbor, map.get(neighbor)+1);
                }else{
                    map.put(neighbor, 1);
                }
            }   
        }
        
        // search for the start point with no parents 
        Queue<DirectedGraphNode> queue = new LinkedList<DirectedGraphNode>();
        ArrayList<DirectedGraphNode> result = new ArrayList<DirectedGraphNode>();
        
        for(DirectedGraphNode node : graph){
            if(!map.containsKey(node)) {
                queue.offer(node);
                // add all nodes without any aprents, since they are all isolated 
                result.add(node);
            }
        }
        
        // topological sort
        while(!queue.isEmpty()){
            DirectedGraphNode cur = queue.poll();
            for(DirectedGraphNode neighbor : cur.neighbors){
                map.put(neighbor, map.get(neighbor)-1);
                if(map.get(neighbor) == 0) {
                    queue.offer(neighbor);
                    result.add(neighbor);
                }
            }
        }
        return result;
    }
}

// method two is DFS

/**
 * Definition for Directed graph.
 * class DirectedGraphNode {
 *     int label;
 *     ArrayList<DirectedGraphNode> neighbors;
 *     DirectedGraphNode(int x) { label = x; neighbors = new ArrayList<DirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */    
    public ArrayList<DirectedGraphNode> topSort(ArrayList<DirectedGraphNode> graph) {
        // write your code here
        HashMap<DirectedGraphNode, Integer> map = new HashMap<DirectedGraphNode, Integer>();
        ArrayList<DirectedGraphNode> result = new ArrayList<DirectedGraphNode>();
        for(DirectedGraphNode node : graph){
            for(DirectedGraphNode nbr : node.neighbors){
                if(map.containsKey(nbr)){
                    map.put(nbr, map.get(nbr)+1);
                }else{
                    map.put(nbr, 1);
                }
            }
        }
        
        for(DirectedGraphNode node : graph){
            if(!map.containsKey(node)) dfs(node, result, map);
        }
        return result;
    }
    
    private void dfs( DirectedGraphNode node, ArrayList<DirectedGraphNode> result, HashMap<DirectedGraphNode, Integer> map){
        
        result.add(node);
        for(DirectedGraphNode n : node.neighbors){
            map.put(n, map.get(n)-1);
            if(map.get(n)==0) dfs(n, result, map);
        }
    }
    
}